File: tyama_icpc2006dB(PKU3007-TJU2602-aizu1142).cpp
#include<iostream>
#include<algorithm>
#include<map>
#include<string>
#define Z(z)m[a+b]=1;m[b+a]=1;reverse(z.begin(),z.end());
using namespace std;main(){string s,a,b;for(cin>>s;cin>>s;){
map<string,int>m;
for(int i=0;++i<s.length();){a=s.substr(0,i),b=s.substr(i);Z(a)Z(b)Z(a)Z(b)}
cout<<m.size()<<endl;
}}

File: tyama_icpc2006dB(PKU3007-TJU2602-aizu1142).java
//090710: Replaced HashMap with HashSet.
import java.io.*;
import java.util.*;
class Main{
  public static void main(String []args){
    Scanner cin=new Scanner(System.in);
    int j=0,n=cin.nextInt();
    cin.nextLine();
    for(;j<n;j++){
    String s=cin.nextLine();
    Set<String> m=new HashSet<String>();
    int i=1;
    for(;i<s.length();i++){
      String s1=s.substring(0,i),s2=s.substring(i),r1=new StringBuilder(s1).reverse().toString(),r2=new StringBuilder(s2).reverse().toString();
      m.add(s1+s2);
      m.add(r1+s2);
      m.add(s1+r2);
      m.add(r1+r2);
      m.add(s2+s1);
      m.add(r2+s1);
      m.add(s2+r1);
      m.add(r2+r1);
    }
    System.out.println(m.size());
    }
  }
}